python/cuml/experimental/explainer/base.pyx

```diff
@@ -0,0 +1,406 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+#
+# Copyright (c) 2020, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import cudf
+import cuml.internals
+import cupy as cp
+import numpy as np
+import pandas
+
+import cuml.common.logger as logger
+from cuml.common.import_utils import has_shap
+from cuml.common.input_utils import input_to_cupy_array
+from cuml.common.logger import warn
+from cuml.experimental.explainer.common import get_dtype_from_model_func
+from cuml.experimental.explainer.common import get_handle_from_cuml_model_func
+from cuml.experimental.explainer.common import get_link_fn_from_str_or_fn
+from cuml.experimental.explainer.common import get_tag_from_model_func
+from cuml.experimental.explainer.common import model_func_call
+from cuml.experimental.explainer.common import output_list_shap_values
+
+from cuml.raft.common.handle cimport handle_t
+from libcpp cimport bool
+from libc.stdint cimport uintptr_t
+
+
+cdef extern from "cuml/explainer/permutation_shap.hpp" namespace "ML":
+
+    void shap_main_effect_dataset "ML::Explainer::shap_main_effect_dataset"(
+        const handle_t& handle,
+        float* dataset,
+        const float* background,
+        int nrows,
+        int ncols,
+        const float* row,
+        int* idx,
+        bool rowMajor) except +
+
+    void shap_main_effect_dataset "ML::Explainer::shap_main_effect_dataset"(
+            const handle_t& handle,
+            double* dataset,
+            const double* background,
+            int nrows,
+            int ncols,
+            const double* row,
+            int* idx,
+            bool rowMajor) except +
+
+
+class SHAPBase():
+    """
+    Base class for SHAP based explainers.
+
+    Parameters
+    ----------
+    model : function
+        Function that takes a matrix of samples (n_samples, n_features) and
+        computes the output for those samples with shape (n_samples). Function
```

Comment:
Ah, bummer so there is no way to use the tags api because we need to take the function rather than the model

Reply:
already use the tags API by getting the owning object of the function (if it exists) and getting tags from that: https://github.com/rapidsai/cuml/blob/39c7262f29ef18000024c8d9741a01179742be01/python/cuml/experimental/explainer/common.py#L21